sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/utils/SerializableRexNode.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.impl.utils;
+
+import java.io.Serializable;
+import org.apache.calcite.rex.RexFieldAccess;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+
+/** SerializableRexNode. */
+public abstract class SerializableRexNode implements Serializable {
```

Comment:
It is confusing to call these `RexNodes`, rename to something like `FieldIndex`?

Reply:
The following work is to add function call support in JOIN condition, e.g. CAST(col) = field access. How to name that then?

I named these new classes to be similar with `RexNode` to preserve some meaning from `RexNode`. Say when see `SerializableRexCall`, it should remind that within this class, there should be a calcite operator with a list of `SerializableRexNode` as operands.

Reply:
I am not sure, I think we will need to think about it a bit more. At this point we're starting to replicate Calcite's relational model representation just because we need to serialize some parts of it.

Also, at the moment this code is used only in `ExtractJoinFields` which doesn't imply function calls. Maybe it's just naming.

Can you share the design for function calls? My concern is that we will need to re-implement that part as well, even though now we have codegen that does similar thing for us in calc rel. Can we re-use that part somehow? E.g. insert a projection which will be backed up by codegen for field accesses and function calls?

Reply:
I can see your concern.

Ok, I will directly work on https://jira.apache.org/jira/browse/BEAM-6112 then.

By doing so, we will have a detailed design/implementation on how to support FunctionCall, ColumnRef and FieldAccess together in Beam.

That should address your concern.